torch/_inductor/cudagraph_trees.py

```diff
@@ -0,0 +1,906 @@
+    is unaffected when the graph is replayed.
+
+    In order to support recording a subsequent cuda graph recording after execution of this graph,
+    we checkpoint the state of the memory pool so that it may later be resumed.
+
+    See [setCheckpointPoolState] for further explanation, as well as
+    https://user-images.githubusercontent.com/13564/222815509-374f3400-f83d-4f7d-8fa6-4a092b3250bb.png
+    """
+
+    def __init__(
+        self,
+        wrapped_function: WrappedFunction,
+        id: GraphID,
+        parent: Optional[CUDAGraphNode],
+        inputs: List[Tensor],
+        cuda_graphs_pool: Tuple[int, int],
+    ):
+        assert isinstance(inputs, (list, tuple))
+
+        self.wrapped_function = wrapped_function
+        self.id = id
+
+        # if this is a root parent will be None. use weakref to prevent reference cycle
+        self._parent = weakref.ref(parent) if parent is not None else None
+        self.cuda_graphs_pool = cuda_graphs_pool
+
+        # A single wrapped function may be recorded multiple times if memory patterns or
+        # invariants change from one execution to the next
+        self.children: Dict[FunctionID, List[CUDAGraphNode]] = defaultdict(list)
+
+        self.device: int = next(
+            (x.device.index for x in inputs if is_cuda_tensor), None
+        )
+
+        # we preserve a single reference to executed outputs that is then referenced
+        # in children to avoid children having to chase parent pointers in the hot path
+        # DO NOT reassign output_weakrefs, only call `clear()`
+        # Path is a series of nodes from root to the current node
+        self.outputs_weakrefs: List[Optional[TensorWeakRef]] = []
+        self.path_weakrefs: List[List[Optional[TensorWeakRef]]] = [
+            node.outputs_weakrefs for node in self._path_from_root
+        ]
+
+        # tensors which are outputs of previous graphs in the tree
+        self.cudagraph_managed_idxs: List[int] = [
+            idx
+            for idx, t in enumerate(inputs)
+            if self._is_cuda_graph_recorded_tensor(t)
+        ]
+
+        self.static_input_idxs: List[int] = list(
+            set(wrapped_function.static_input_idxs) | set(self.cudagraph_managed_idxs)
+        )
+
+        self.static_input_data_ptrs: List[int] = [
+            (inputs[i].data_ptr() if i in self.static_input_idxs else None)
+            for i in range(len(inputs))
+        ]
+        self.expanded_dims: List[List[int]] = [
+            get_expanded_dims(x) if idx not in self.static_input_idxs else []
+            for idx, x in enumerate(inputs)
+        ]
+
+        # For each node in path, which outputs were observed to be live
+        # before invoking graph recording, and after graph recording
+        self.recorded_liveness_before_graph: List[List[bool]] = []
+        self.recorded_liveness_after_graph: List[List[bool]] = []
+
+        # indices into node at level i for output j
```

Comment:
I have no idea, at thisnpoint, what a "level" is

Reply:
lol took that terminology from wikipedia. the level of the tree is distance from the root. maybe depth is more idiomatic. or i'll just spell out with the full description..

Reply:
I agree level is unambiguous for this. It's surprising though because you don't have a root, you would traverse parent pointers to populate this.

Reply:
It might be cleaner for me to pass in a root, instead of a parent. it doesn't matter perf wise (dont traverse on hot path) but it would be cleaner perhaps